MoreLinq/Batch.cs

```diff
@@ -87,69 +86,65 @@ public static IEnumerable<IEnumerable<TSource>> Batch<TSource>(this IEnumerable<
+                    {
+                        yield break;
+                    }
+                    case IReadOnlyList<TSource> list when list.Count <= size:
                     {
                         var bucket = new TSource[list.Count];
                         for (var i = 0; i < list.Count; i++)
                             bucket[i] = list[i];
                         yield return resultSelector(bucket);
+                        yield break;
                     }
-                }
-                case IReadOnlyCollection<TSource> collection when collection.Count <= size:
-                {
-                    return Batch(collection.Count);
-                }
-                default:
-                {
-                    return Batch(size);
-                }
-
-                IEnumerable<TResult> Batch(int size)
-                {
-                    TSource[]? bucket = null;
-                    var count = 0;
-
-                    foreach (var item in source)
+                    case IReadOnlyCollection<TSource> collection when collection.Count <= size:
+                    {
+                        size = collection.Count;
+                        goto default;
+                    }
+                    default:
                     {
-                        bucket ??= new TSource[size];
-                        bucket[count++] = item;
+                        TSource[]? bucket = null;
+                        var count = 0;
 
-                        // The bucket is fully buffered before it's yielded
-                        if (count != size)
-                            continue;
+                        foreach (var item in source)
+                        {
+                            bucket ??= new TSource[size];
+                            bucket[count++] = item;
 
-                        yield return resultSelector(bucket);
+                            // The bucket is fully buffered before it's yielded
+                            if (count != size)
+                                continue;
 
-                        bucket = null;
-                        count = 0;
-                    }
+                            yield return resultSelector(bucket);
 
-                    // Return the last bucket with all remaining elements
-                    if (count > 0)
-                    {
-                        Array.Resize(ref bucket, count);
-                        yield return resultSelector(bucket);
+                            bucket = null;
+                            count = 0;
+                        }
+
+                        // Return the last bucket with all remaining elements
+                        if (count > 0)
+                        {
+                            Array.Resize(ref bucket, count);
+                            yield return resultSelector(bucket);
+                        }
+                        break;
```

Comment:
nit: blank line for clarity:

```suggestion

                        break;
```